radix-engine/src/kernel/call_frame.rs

```diff
@@ -100,12 +147,38 @@ impl NodeVisibility {
         self.0.iter().any(|x| x.is_normal())
     }
 
-    pub fn can_be_reference_copied_to_frame(&self) -> Option<StableReferenceType> {
+    pub fn is_global(&self) -> bool {
         for v in &self.0 {
-            if let Visibility::StableReference(t) = v {
-                return Some(t.clone());
+            if let Visibility::StableReference(StableReferenceType::Global) = v {
+                return true;
             }
         }
+        return false;
+    }
+
+    pub fn reference_origin(&self, node_id: NodeId) -> Option<ReferenceOrigin> {
+        let mut found_direct_access = false;
+        for v in &self.0 {
+            match v {
+                Visibility::StableReference(StableReferenceType::Global) => {
+                    return Some(ReferenceOrigin::Global(GlobalAddress::new_or_panic(
+                        node_id.0,
+                    )));
+                }
+                Visibility::StableReference(StableReferenceType::DirectAccess) => {
+                    found_direct_access = true
```

Comment:
Can we add a TODO here that we'd like to change the return to `Vec<ReferenceOrigin>`?

This "global supersedes direct access reference" is very arbitrary and unsound.